### PR TITLE
Mainnet

### PR DESCRIPTION
**Pull request by bxmmm1**

### CODE REVIEW

File: src/PufferVaultMainnet.sol [hunk 1]
```diff
@@ -46,22 +64,29 @@ contract PufferVaultMainnet is PufferVault {
      * Sum of EL assets + Vault Assets
      */
     function totalAssets() public view virtual override returns (uint256) {
+        // If we are dealing with native ETH deposit, we need to deduct callvalue from the balance
+        uint256 callValue;
+        assembly {
+            callValue := callvalue()
```
Comment (shayanb) [inline src/PufferVaultMainnet.sol:70]: Need more insights in what `callvalue()` is and why this implementation

File: src/PufferVaultMainnet.sol [hunk 2]
```diff
@@ -80,20 +105,107 @@ contract PufferVaultMainnet is PufferVault {
...
         return assets;
     }
 
-    // slither-disable-next-line dead-code
-    function _wrap(uint256 amount) internal {
-        _WETH.deposit{ value: amount }();
+    /**
+     * @param withdrawalAmount is the assets amount, not shares
+     */
+    function _checkDailyWithdrawalLimits(uint256 withdrawalAmount) internal {
```
Comment (shayanb) [inline src/PufferVaultMainnet.sol:122]: function name needs to reflect the functionality -- updates state variables. suggestion: checkandUpdate...

File: src/PufferVaultMainnet.sol [hunk 3]
```diff
@@ -80,20 +105,107 @@ contract PufferVaultMainnet is PufferVault {
...
+        VaultStorage storage $ = _getPufferVaultStorage();
+
+        // If daily withdrawal limit is 0, then there is no limit
+        if ($.dailyWithdrawalLimit == 0) {
```
Comment (shayanb) [inline src/PufferVaultMainnet.sol:126]: dailyWithdrawalLimit should be maxUInt instead for no limit case. (dailyWithdrawalLimit == maxUint --> return ) 

(maybe set `dailyWithdrawalLimit` in initializer to 0 for explicit declaration)

File: src/PufferVaultMainnet.sol [hunk 4]
```diff
@@ -80,20 +105,107 @@ contract PufferVaultMainnet is PufferVault {
         }
 
         uint256 assets = previewRedeem(shares);
+
+        _checkDailyWithdrawalLimits(assets); //@todo figure if it is better to override `maxWithdraw`
```
Comment (shayanb) [inline src/PufferVaultMainnet.sol:109]: doubt it's necessary to update maxWithdraw as it is global limit. Any edge cases you see it might be needed?

File: src/PufferVaultMainnet.sol [hunk 5]
```diff
@@ -80,20 +105,107 @@ contract PufferVaultMainnet is PufferVault {
...
+     */
+    function transferETH(address to, uint256 ethAmount) external restricted {
+        // Our Vault will hold ETH & WETH
+        // If we don't have enough ETH for the transfer, unwrap WETH
+        uint256 ethBalance = address(this).balance;
+        if (ethBalance < ethAmount) {
+            _WETH.withdraw(ethAmount - ethBalance);
+        }
+
+        /// @solidity memory-safe-assembly
```
Comment (shayanb) [inline src/PufferVaultMainnet.sol:175]: Suggestion to use normal .call() to transfer the amount

The assembly looks out of the place as we don't have that much assembly, and this function is not used as often. open to discuss

File: src/PufferVaultMainnet.sol [hunk 6]
```diff
@@ -80,20 +105,107 @@ contract PufferVaultMainnet is PufferVault {
...
+                revert(0x1c, 0x04)
+            }
+        }
+    }
+
+    /**
+     * @notice Allows the `msg.sender` to burn his shares
+     * @param shares The amount of shares to burn
+     */
+    function burn(uint256 shares) public {
```
Comment (shayanb) [inline src/PufferVaultMainnet.sol:188]: Add `restricted` modifier.

File: test/Integration/PufferTest.integration.t.sol [hunk 7]
```diff
@@ -24,6 +24,7 @@ import { ILidoWithdrawalQueue } from "src/interface/Lido/ILidoWithdrawalQueue.so
 import { IEigenLayer } from "src/interface/EigenLayer/IEigenLayer.sol";
 import { IStrategy } from "src/interface/EigenLayer/IStrategy.sol";
 import { Timelock } from "src/Timelock.sol";
+import { IWETH } from "src/interface/Other/IWETH.sol";
```
Comment (shayanb) [inline test/Integration/PufferTest.integration.t.sol:27]: Todo: integration testing with Weth, flows such as deposit, withdraw, redeem, etc

as well as some edge cases such as:
- withdraw too much
- different balances (weth, eth) v.s. diff withdrawalAmounts
- withdrawalLimits changes

File: src/PufferVaultMainnet.sol (hunk 1 above)
Comment (bxmmm1) [reply src/PufferVaultMainnet.sol:70]: I've added a comment that explains this, callValue will only be populated in `depositETH` flow.

File: src/PufferVaultMainnet.sol (hunk 2 above)
Comment (bxmmm1) [reply src/PufferVaultMainnet.sol]: Changed the logic for `maxDeposit` and `maxRedeem` so that we are compliant with the 4626 standard.

File: src/PufferVaultMainnet.sol (hunk 1 above)
Comment (shayanb) [reply src/PufferVaultMainnet.sol:70]: thanks that answered all my initial questions. 

I'm wondering if this could break anything with defi integrations. Imagine a scenario that this minting/burning is happening in a chain of internal transactions (within one transaction), would this substraction create a difference between the amount minted/burned if callvalue()/msg.value is not intended for puffer contract?

File: src/PufferVaultMainnet.sol (hunk 3 above)
Comment (shayanb) [reply src/PufferVaultMainnet.sol]: I think this is also fixed by changing the logic to use default 4626 limits, I'll resolve this after reviewing the new flow.

File: src/PufferVaultMainnet.sol (hunk 1 above)
Comment (shayanb) [reply src/PufferVaultMainnet.sol:70]: Update: seems to be a non-issue. just make sure we have tests for this

File: src/PufferVaultMainnet.sol [hunk 8]
```diff
@@ -7,61 +7,110 @@ import { ILidoWithdrawalQueue } from "src/interface/Lido/ILidoWithdrawalQueue.so
...
-    //@todo weth wrapping and unwrapping logic, native ETH deposit method
+    /**
+     * @notice Calculates the maximum amount of assets that can be withdrawn by the `owner`.
+     * @dev This function considers both the remaining daily withdrawal limit and the `owner`'s balance.
+     * @param owner The address of the owner for which the maximum withdrawal amount is calculated.
+     * @return maxAssets The maximum amount of assets that can be withdrawn by the `owner`.
+     */
+    function maxWithdraw(address owner) public view virtual override returns (uint256 maxAssets) {
+        uint256 remainingAssets = getRemainingAssetsDailyWithdrawalLimit();
+        uint256 maxUserAssets = _convertToAssets(balanceOf(owner), Math.Rounding.Floor);
```
Comment (shayanb) [inline src/PufferVaultMainnet.sol:84]: Use `previewRedeem()` for consistency with the rest of the code.

File: src/PufferVaultMainnet.sol [hunk 9]
```diff
@@ -7,61 +7,110 @@ import { ILidoWithdrawalQueue } from "src/interface/Lido/ILidoWithdrawalQueue.so
...
      */
     function initialize() public reinitializer(2) {
-        ERC4626Storage storage $ = _getERC4626StorageInternal();
-        $._asset = _WETH;
+        // In this initialization, we swap out the underlying stETH with WETH
+        ERC4626Storage storage erc4626Storage = _getERC4626StorageInternal();
+        erc4626Storage._asset = _WETH;
+
+        VaultStorage storage $ = _getPufferVaultStorage();
+        $.lastWithdrawalDay = uint64(block.timestamp / 1 days);
```
Comment (shayanb) [inline src/PufferVaultMainnet.sol:51]: The daily limit needs to be set to 0, as in withdrawal is disabled. 
(We need to explicitly set that to 0 here for readability and auditors.)

File: src/PufferVaultMainnet.sol [hunk 10]
```diff
@@ -80,20 +129,100 @@ contract PufferVaultMainnet is PufferVault {
...
+     */
+    function setDailyWithdrawalLimit(uint96 newLimit) external restricted {
+        VaultStorage storage $ = _getPufferVaultStorage();
+        emit DailyWithdrawalLimitSet($.dailyAssetsWithdrawalLimit, newLimit);
+        $.dailyAssetsWithdrawalLimit = newLimit;
+    }
+
+    function getRemainingAssetsDailyWithdrawalLimit() public view virtual returns (uint96) {
+        VaultStorage storage $ = _getPufferVaultStorage();
+        return $.dailyAssetsWithdrawalLimit - $.assetsWithdrawnToday;
```
Comment (shayanb) [inline src/PufferVaultMainnet.sol:202]: In the scenario that the limit is not 0 and is set to 0, then this would be negative and revert, resulting in other functions failing. 
add a check 
- $.dailyAssetsWithdrawalLimit < $.assetsWithdrawnToday return 0
- else ...

File: script/UpgradePuffETH.s.sol [hunk 11]
```diff
@@ -0,0 +1,97 @@
+// SPDX-License-Identifier: GPL-3.0
```
Comment (CheyenneAtapour) [inline script/UpgradePuffETH.s.sol:1]: @bxmmm1 is there a test that shows this upgrade works as expected? E.g. we can call the upgrade on the vault and try calling some of the functions in a test

Comment (bxmmm1) [reply script/UpgradePuffETH.s.sol:1]: https://github.com/PufferFinance/pufETH/blob/feature/mainnet/test/Integration/PufferTest.integration.t.sol#L362

**Comment by JasonVranek** [comment]
LGTM, my main suggestions are to drop the swapping logic and keep the stETH redemption logic from V1 contracts.
